### PR TITLE
backend-common: lock down UrlReader to only read from allowed URLs

### PR DESCRIPTION
**Pull request by Rugvip**
The catalog placeholder processor needs to be locked down a bit more. Figuring the best place to do that is to apply it across all URL reading, as that provides a way to overall control where the backend is allowed to read from.

To be able to read data from for example `example.com`, you now need to add the following config:

```
backend:
  reading:
    allow:
      - host: example.com
```

### CODE REVIEW

File: docs/features/software-catalog/descriptor-format.md [hunk 1]
```diff
@@ -131,6 +131,19 @@ spec:
     $text: https://petstore.swagger.io/v2/swagger.json
 ```
 
+Note that to be able to read from targets that are outside of the normal
+integrations points such as `github.com`, you'll need to explicitly allow it by
```
Comment (freben) [inline docs/features/software-catalog/descriptor-format.md:135]:
```suggestion
integration points such as `github.com`, you'll need to explicitly allow it by
```

File: packages/backend-common/config.d.ts [hunk 2]
```diff
@@ -94,6 +94,25 @@ export interface Config {
       optionsSuccessStatus?: number;
     };
 
+    /**
+     * Configuration relates to URL reading, used for example for reading catalog info
```
Comment (freben) [inline packages/backend-common/config.d.ts:98]:
```suggestion
     * Configuration related to URL reading, used for example for reading catalog info
```

File: packages/backend-common/config.d.ts [hunk 3]
```diff
@@ -94,6 +94,25 @@ export interface Config {
...
+     * files, scaffolder templates, and techdocs content.
+     */
+    reading?: {
+      /**
+       * A list of targets to allow outgoing requests to. Users will be able to make
+       * request on behalf of backend to the targets that are allowed by this list.
```
Comment (freben) [inline packages/backend-common/config.d.ts:104]:
```suggestion
       * requests on behalf of the backend to the targets that are allowed by this list.
```

File: packages/backend-common/src/reading/FetchUrlReader.test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      logger: getVoidLogger(),
+      treeResponseFactory: ReadTreeResponseFactory.create({
+        config: new ConfigReader({}),
+      }),
+    });
+
+    expect(entries.length).toBe(1);
+    const [{ predicate }] = entries;
+
+    expect(predicate(new URL('https://example.com/test'))).toBe(true);
```
Comment (freben) [inline packages/backend-common/src/reading/FetchUrlReader.test.ts:51]: also check `a.example.com` to be false

File: packages/backend-common/src/reading/FetchUrlReader.ts [hunk 5]
```diff
@@ -16,12 +16,39 @@
...
-import { ReadTreeResponse, UrlReader } from './types';
+import { ReaderFactory, ReadTreeResponse, UrlReader } from './types';
 
 /**
  * A UrlReader that does a plain fetch of the URL.
  */
 export class FetchUrlReader implements UrlReader {
+  /**
+   * The factory creates a single reader that will be used for reading any URL that's listed
+   * in configuration at `backend.reading.allow`. The allow list contains a list of object describing
```
Comment (freben) [inline packages/backend-common/src/reading/FetchUrlReader.ts:27]:
```suggestion
   * in configuration at `backend.reading.allow`. The allow list contains a list of objects describing
```

File: packages/backend-common/src/reading/FetchUrlReader.ts [hunk 6]
```diff
@@ -16,12 +16,39 @@
...
+   *   Either full hostnames to match, or subdomain wildcard matchers with a leading `*`.
+   *   For example `example.com` and `*.example.com` are valid values, `prod.*.example.com` is not.
+   */
+  static factory: ReaderFactory = ({ config }) => {
+    const predicates =
+      config
+        .getOptionalConfigArray('backend.reading.allow')
+        ?.map(allowConfig => {
+          const host = allowConfig.getString('host');
+          if (host.startsWith('*') && host.length > 1) {
```
Comment (freben) [inline packages/backend-common/src/reading/FetchUrlReader.ts:40]:
```suggestion
          if (host.startsWith('*.') && host.length > 1) {
```

File: plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts [hunk 7]
```diff
@@ -57,7 +62,12 @@ describe('UrlReaderProcessor', () => {
 
   it('should fail load from url with error', async () => {
     const logger = getVoidLogger();
-    const reader = UrlReaders.default({ logger, config: new ConfigReader({}) });
+    const reader = UrlReaders.default({
+      logger,
+      config: new ConfigReader({
+        backend: { reading: { allow: [{ host: 'localhost' }] } },
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts:68]: For local testing, is there any need to add explicitly (or in the default app-config) typical localhost addresses / IPs? Obv not `*.local`, but `localhost` et al

**Review by freben — APPROVED** [review]

File: app-config.yaml [hunk 8]
```diff
@@ -16,6 +16,10 @@ backend:
     credentials: true
   csp:
     connect-src: ["'self'", 'http:', 'https:']
+  reading:
+    allow:
+      - host: example.com
+      - host: '*.mozilla.org'
```
Comment (Fox32) [inline app-config.yaml:22]: I think we should have `github.com` in the default configuration, as we have some API entities that are using this feature: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/spotify-api.yaml

**Comment by Fox32** [comment]
I wonder if this break the behavior described here: https://github.com/backstage/backstage/pull/4033#discussion_r555919725

File: app-config.yaml (hunk 8 above)
Comment (freben) [reply app-config.yaml:22]: As long as the default readers are added as well (`GithubUrlReader` for example), they automatically add a config entry for the public site whether you actually have it in your config or not. So if I understand things correctly it'll still work, just not fall down to the `FetchUrlReader`.

Comment (Fox32) [reply app-config.yaml:22]: Ah, that make sense

**Comment by Rugvip** [comment]
@Fox32 File locations won't be using the URL reader at all, so that's np. In general I'm expecting this to only have an impact on usage of placeholders in catalog info files, as catalog info files are pretty much loaded via repo integrations rather than the general URL reading.

**Comment by Rugvip** [comment]
Btw, this doesn't lock down ports at all atm, thoughts on that? I'm thinking we either keep it open and it's up to the infra to make sure that sensitive ports aren't exposed on surrounding services. Or we hardcode it to 80/443 for now and possibly add some config keys to allow other ports in the future.

File: plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts (hunk 7 above)
Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts:68]: Not a very common thing to do I think?

**Comment by Rugvip** [comment]
Handling ports by simply including them in the check, so now you'd have to allow e.g. `example.com:8080`

File: plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts (hunk 7 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.test.ts:68]: Not sure!

**Comment by freben** [comment]
That probably makes sense. A bit more inconvenient than saying `http://` or `https://` at the start though since it'll be so overwhelmingly common? I know we can't verify that `https://something:1234` then actually talks https, at least not in the predicate check phase, but ...

**Comment by freben** [comment]
ah i see now that it's actually covered anyway

**Comment by Fox32** [comment]
> @Fox32 File locations won't be using the URL reader at all...

Thanks, good to know

**Comment by Rugvip** [comment]
@Fox32 It was considered, but I think we rightfully concluded that the security implications of that are a bit too scary x)

**Pull request merged**